source/sessions/tests/driver-sessions-dirty-session-errors.yml

```diff
@@ -0,0 +1,353 @@
+        expectResult: []
+      - name: assertDifferentLsidOnLastTwoCommands
+        object: testRunner
+        arguments:
+          client: *client0
+    expectEvents:
+      - client: *client0
+        events:
+          - commandStartedEvent: &insert_attempt
+              command:
+                insert: *collection0Name
+                documents: 
+                  - { _id: 2 }
+                ordered: true
+                lsid: { $$sessionLsid: *session0 }
+                txnNumber: 1
+              commandName: insert
+              databaseName: *database0Name
+          - commandStartedEvent: *insert_attempt
+          - commandStartedEvent:
+              command:
+                insert: *collection0Name
+                documents: 
+                  - { _id: 3 }
+                ordered: true
+                lsid: { $$sessionLsid: *session0 }
+                txnNumber: 2
+              commandName: insert
+              databaseName: *database0Name
+          - commandStartedEvent: &find_with_implicit_session_event
+              command:
+                find: *collection0Name
+                filter: { _id: -1 }
+                # There is no explicit session to use with $$sessionLsid, so
+                # just assert an arbitrary lsid document
+                lsid: { $$type: object }
+              commandName: find
+              databaseName: *database0Name
+    outcome:
+      - collectionName: *collection0Name
+        databaseName: *database0Name
+        documents:
+          - { _id: 1 }
+          - { _id: 2 }
+          - { _id: 3 }
+
+  - description: "Dirty explicit session is discarded (findAndModify)"
+    operations:
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *client0
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 1 }
+            data:
+              failCommands: [ findAndModify ]
+              closeConnection: true
+      - name: assertSessionNotDirty
+        object: testRunner
+        arguments:
+          session: *session0
+      - name: findOneAndUpdate
+        object: *collection0
+        arguments:
+          session: *session0
+          filter: { _id: 1 }
+          update: { $inc: { x: 1 } }
+          returnDocument: Before
+        expectResult: { _id: 1 }
+      - name: assertSessionDirty
+        object: testRunner
+        arguments:
+          session: *session0
+      - name: endSession
+        object: *session0
+      - *find_with_implicit_session
+      - name: assertDifferentLsidOnLastTwoCommands
+        object: testRunner
+        arguments:
+          client: *client0
+    expectEvents:
+      - client: *client0
+        events:
+          - commandStartedEvent: &findAndModify_attempt
+              command:
+                findAndModify: *collection0Name
+                query: { _id: 1 }
+                update: { $inc: { x: 1 } }
+                new: false
```

Comment:
it appears that libmongoc's `mongoc_collection_find_and_modify_with_opts` only sends `new` in a command when it is set to `true`, presumably since `false` is the server default:
https://github.com/mongodb/mongo-c-driver/blob/28c2ca827d318cb3ff3c832982e65b39408707c2/src/libmongoc/src/mongoc/mongoc-collection.c#L3335-L3337

perhaps this should use `$$unsetOrMatches`? or should we consider this a libmongoc issue since it makes libmongoc less resilient to the server hypothetically changing the default?

Reply:
Thanks for point this out. I took at look at existing specs and noted that some legacy transaction spec tests (same format used by legacy session tests) also use `new: false`. libmongoc purposely ignores assertions on `new` fields in commandStartedEvents for `findAndModify` (see: https://github.com/mongodb/mongo-c-driver/blob/1.21.0/src/libmongoc/tests/json-test-monitoring.c#L418); however, there's probably a small bug there since it completely ignores the field instead of just a `new: false` assertion. I don't think it's worth reporting for a legacy test runner, though (cc: @kevinAlbs).

This came up previously in some unified specs and I see that I used `$$unsetOrMatches` there for reasons cited in https://github.com/mongodb/specifications/pull/980/files#r635414676. I see no reason we can't do the same here.

Reply:
SGTM